src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,411 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Clustering;
+using Hazelcast.Core;
+using Hazelcast.DistributedObjects;
+using Hazelcast.Exceptions;
+using Hazelcast.Protocol;
+using Hazelcast.Protocol.Models;
+
+namespace Hazelcast.CP
+{
+    /// <summary>
+    /// Implementation of <see cref="IFencedLock"/>
+    /// </summary>
+    internal partial class FencedLock : CPDistributedObjectBase, IFencedLock
+    {
+        /// <summary>
+        /// Gets the current context identifier.
+        /// </summary>
+        /// <remarks>
+        /// Hazelcast APIs call this the thread identified and maintain locks "per threads",
+        /// so we are keeping the name here internally, but in reality this is not a thread
+        /// identifier anymore - it is attached to the async context so it can flow with
+        /// async operations.
+        /// </remarks>
+        private static long ContextId => AsyncContext.Current.Id;
+        private readonly Dictionary<long, long> _lockedThreadToSession = new Dictionary<long, long>();
+        private readonly CpSubsystemSession _subsystemSession;
+        private readonly CPGroupId _groupId;
+
+        public const long InvalidFence = 0;
+        ICPGroupId IFencedLock.CPGroupId => _groupId;
+        long IFencedLock.InvalidFence => InvalidFence;
+
+        public FencedLock(CpSubsystemSession subsystemSession, string name, CPGroupId groupId, Cluster cluster) : base(ServiceNames.FencedLock, name, groupId, cluster)
+        {
+            _subsystemSession = subsystemSession;
+            _groupId = groupId;
+        }
+
+        #region IFencedLock Methods
+        /// <inheritdoc/>  
+        public async Task<long> GetFenceAsync()
+        {
+            long threadId = ContextId;
+            long sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            if (sessionId == CpSubsystemSession.NoSessionId)
+            {
+                _lockedThreadToSession.Remove(threadId);
+                throw new SynchronizationLockException();
+            }
+
+            var ownership = await RequestLockOwnershipStateAsync().CfAwait();
+
+            if (ownership.LockedBy(threadId, sessionId))
+            {
+                _lockedThreadToSession[threadId] = sessionId;
+                return ownership.Fence;
+            }
+
+            VerifyNoLockedSessionExist(threadId);
```

Comment:
Here, and in many places in this class. `VerifyNoLockedSessionExist` ends up reading and writing `_lockedThreadToSession` and this dictionary is not concurrent and I am not sure that this is all thread-safe?

Reply:
You are right, it is not. I also realized the same thing during testing and fixed to concurrent.